### ISSUE TITLE
[1860] Upgrade hex availability miscalculation

### ISSUE THREAD

**Issue by Blobbley** [tobymao/18xx#4029]
https://18xx.games/game/25057?action=366

Here BHIR can't Upgrade Ryde because with the already built track it would require a 4 train to reach it. However, I could reach Ryde with a 2 train AFTER it had upgraded.

At the time I thought the implementation was adhering to the rules and I shouldn't be able to upgrade Ryde (you can only upgrade tiles you can already access - and reading the rules this seems to be the letter of them) but that would mean in this game (BHIR again):

https://18xx.games/game/26387?action=412

BHIR would not be allowed to upgrade Sanddown and therefore could never have a legal route. Which is definitely not in my interpretation of the spirit of the rules.

TLDR: in ID 25057 BHIR has a 2 possible routes to Ryde and owns a 3+ train. One route already exists and requires a 4 train which prohibits the upgrade, but a route could be created by an upgrade that would only require a 2 train. Can upgrades *create* routes, or can they only be made on tiles that already have routes existing to them. The latter seems to be the letter of the rules but in some cases would make certain situations untenable.

**Comment by roseundy** [comment]
I agree with you on both counts: I think BHIR should be able to upgrade Ryde and I think the rules are very unclear. I think I'm going to post this on BGG and see if I can get a ruling from Mike Hutton.

**Comment by roseundy** [comment]
Actually, this is a plain old bug on this line: https://github.com/tobymao/18xx/blob/f264ba38977daddf3105d99b6fa366316db0ffb1/lib/engine/game/g_1860/step/track.rb#L120

It found a path on the existing tile that was out of reach and stopped looking :-/
